### PR TITLE
feat(spanner): support custom field type

### LINKED ISSUE #854: Spanner: Support Custom Marsheling and Unmarsheling
It would be great to have the possibility to have custom un-/marsheling for custom types support a la JSON de-&encoder or CSV de-&encoder:

Example:

DDL:
```sql
CREATE TABLE ProductInfo (
    ProductID    INT64 NOT NULL,
    UpdateDate   TIMESTAMP NOT NULL,
    Data         STRING
) PRIMARY KEY (ProductID)
```

```go
type customField struct {
  Prefix string
  Suffix string
}
type dbProductInfo struct {
    ProductID  int64       `spanner:"ProductID"`
    UpdateDate time.Time   `spanner:"UpdateDate"`
    Data       customField `spanner:"Data"`
}

// Convert the customField 
func (cf *customField) MarshalSpanner() (string, error) {
    var b bytes.Buffer
    b.WriteString(cf.Prefix)
    b.WriteString("-")
    b.WriteString(cf.Suffix)
    return b.String(), nil
}

// Convert Spanner string to customField
func (cf *customField) UnmarshalSpanner(val string) (err error) {
   s := strings.SplitAfter(val, "-")
   if len(s) > 1 {
       cf.Prefix = s[0]
       cf.Suffix = s[1]
   }
   return nil
}
```

### PR DESCRIPTION
**Pull request by hengfengli**
Fixes #854

### CODE REVIEW

**Comment by hengfengli** [comment]
The following example works fine: 

```go
package main

import (
	"bytes"
	"context"
	"fmt"
	"log"
	"strings"
	"time"

	"cloud.google.com/go/spanner"
	"google.golang.org/api/iterator"
)

type customField struct {
	Prefix string
	Suffix string
}

// Convert the customField
func (cf customField) EncodeSpanner() (interface{}, error) {
	var b bytes.Buffer
	b.WriteString(cf.Prefix)
	b.WriteString("-")
	b.WriteString(cf.Suffix)
	return b.String(), nil
}

// Convert Spanner string to customField
func (cf *customField) DecodeSpanner(val interface{}) (err error) {
	strVal, ok := val.(string)
	if !ok {
		return fmt.Errorf("failed to decode customField: %v", val)
	}
	s := strings.Split(strVal, "-")
	if len(s) > 1 {
		cf.Prefix = s[0]
		cf.Suffix = s[1]
	}
	return nil
}

type dbProductInfo struct {
	ProductID  int64       `spanner:"ProductID"`
	UpdateDate time.Time   `spanner:"UpdateDate"`
	Data       customField `spanner:"Data"`
}

func main() {
	ctx := context.Background()
	client, err := spanner.NewClient(ctx, "[DB_PATH]")
	if err != nil {
		log.Fatal(err)
	}
	defer client.Close()

	// Write

	// d := dbProductInfo{
	// 	ProductID:  1234,
	// 	UpdateDate: time.Now(),
	// 	Data:       customField{"prefix", "suffix"},
	// }

	// m, err := spanner.InsertStruct("ProductInfo", &d)
	// if err != nil {
	// 	log.Fatalf("err: %v", err)
	// }

	// ms := []*spanner.Mutation{m}
	// _, err = client.Apply(ctx, ms)
	// if err != nil {
	// 	log.Fatalf("err: %v", err)
	// }

	// Read

	stmt := spanner.Statement{SQL: "SELECT * FROM ProductInfo"}
	iter := client.Single().Query(ctx, stmt)
	defer iter.Stop()

	for {
		row, err := iter.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			log.Fatal(err)
		}

		var d dbProductInfo
		if err := row.ToStruct(&d); err != nil {
			log.Fatal(err)
		}
		fmt.Println(d)
	}
}
```

**Comment by hengfengli** [comment]
@olavloite The only issue I have is that if the custom field is a struct we have to implement the interface via: 

```go
type dbProductInfo struct {
        ...
	Data       customField `spanner:"Data"`
}

func (cf customField) EncodeSpanner() (interface{}, error) {
    ...
}

func (cf *customField) DecodeSpanner(val interface{}) (err error) {
    ...
}
```

As you can see `EncodeSpanner` is implemented on a value receiver and `DecodeSpanner` is implemented on a pointer receiver. 

The reason why we can't implement `EncodeSpanner` on a pointer receiver is that if we do the following: 

```go
func (cf *customField) EncodeSpanner() (interface{}, error) {
    ...
}
```

Then, `v.(Encoder)` would fail: 

```go
if encodedVal, ok := v.(Encoder); ok {    // ok => false 
}
```

Basically, if a method of an interface (assume only one method) is implemented on a pointer receiver, the struct cannot be casted to the interface (it doesn't think the struct has implemented the interface). 

Similar examples in: 

https://github.com/googleapis/google-cloud-go/blob/e143982c0485e1ff2726f7e2b607c26997dd0d90/spanner/value.go#L78-L103

The following example should demonstrate the issue: 

```go
package main

import (
	"fmt"
	"math"
)

type Vertex struct {
	X, Y float64
}

type M interface {
	Abs() float64
}

func (v *Vertex) Abs() float64 {
	return math.Sqrt(v.X*v.X + v.Y*v.Y)
}

func t(v interface{}) {
	_, ok := v.(M)
	fmt.Println("ok:", ok)
}

func main() {
	v := Vertex{3, 4}

	t(v)
}
```

**Comment by olavloite** [comment]
> Have you met this issue before?

Hmm... No, I was not really aware of that, but the reference to how its done in `value.go` does indicate that this is to be expected. I guess that if we document it properly with an example it should not really be a problem for the users.

File: spanner/value.go [hunk 1]
```diff
@@ -50,6 +50,18 @@ var (
 	jsonNullBytes = []byte("null")
 )
 
+// Encoder is the interface implemented by a custom type that can be encoded to
+// a supported type by Spanner.
```
Comment (olavloite) [inline spanner/value.go:54]: Could we include a small example in the documentation here to show how this should be used, especially considering the necessity to implement one of the interface methods using a value and the other using a pointer.

Comment (hengfengli) [reply spanner/value.go]: I have added code examples to before Encoder and Decoder. Please take a look.

**Comment by hengfengli** [comment]
@olavloite , according to your feedback, I have added end-to-end tests for both encoding and decoding and also code examples in the comments. 

One thing I realize during writing tests is that currently we don't support arrays of custom fields yet. Do you think it is necessary to implement them in this PR? E.g., 

```go
type dbProductInfo struct {
	ProductID  int64         `spanner:"ProductID"`
	UpdateDate time.Time     `spanner:"UpdateDate"`
	Data       []customField `spanner:"Data"`
}
```

**Pull request merged**